### PR TITLE
Updating generic math to support user-defined checked operators

### PR DESCRIPTION
**Pull request by tannergooding**

### CODE REVIEW

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (tannergooding) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:727]: Changes were automated via https://github.com/dotnet/arcade/pull/8769, removing the need to continuously "manually" update for static abstracts

File: src/libraries/System.Private.CoreLib/src/System/Byte.cs
Comment (deeprobin) [inline src/libraries/System.Private.CoreLib/src/System/Byte.cs:304]: Shouldn't we specify visibility here?
Or is that not good practice for interface implementations?

I haven't worked with interface operators yet :D.

File: src/libraries/System.Private.CoreLib/src/System/Byte.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Byte.cs:395]: What values of left and right would this throw an overflow exception for?

File: src/libraries/System.Runtime/tests/System/ByteTests.GenericMath.cs
Comment (deeprobin) [inline src/libraries/System.Runtime/tests/System/ByteTests.GenericMath.cs:59]: Imo, the whole thing would be a bit more structured if you made the Equality Checks a Theory with InlineData and the Throws Check a separate Fact.

**Review by deeprobin — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1515]: This is interesting.  DateTime.MaxValue + TimeSpan.FromMilliseconds(1) will throw an exception, regardless of checked.  In a generic math context, that feels a little strange, but I'm not sure there's a better answer (as long as we continue to want these only-slightly-math-related types implementing these interfaces).

File: src/libraries/System.Private.CoreLib/src/System/Byte.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Byte.cs:395]: For .NET/IL, Division for the primitives doesn't overflow and in fact there is no `div.ovf` instruction in IL. The behavior is that division by `zero` always throws a `DivideByZeroException` and that `MinValue / -1` always throws `OverflowException`.

For C#, the language formally defines `checked(x / y)` to throw `ArithmeticException` (or a subclass) for `MinValue / -1`. While it leaves `unchecked(x / y)` for the same to be "implementation-defined` on whether it throws or goes unreported and returns `x`.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1557]: I assume we're only implementing these because we have to based on the interface hierarchy, and if we had default interface implementations, we would just let the base virtual implementations delegate to the non-checked?

File: src/libraries/System.Private.CoreLib/src/System/Byte.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Byte.cs:395]: But byte can't be -1.  So is the checked on the rhs here doing anything?  Seems like you get identical behavior / asm with or without it, yes?

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1515]: I think this is an area where we'll want to consider the implications of there now being `user-defined checked operators` for certain core types (like `DateTime` and similar).

One could imagine us "properly" supporting checked/unchecked contexts for these types now. One could also imagine us continuing to always throw on overflow.

`checked(expr)` doesn't necessarily mean that `expr` will throw on overflow; and `unchecked(expr)` doesn't necessarily mean that `expr` will not throw on overflow. It really depends on the type and what operations it decides to support.

We'll probably want to decide on guidance as well, including the implications of taking the behavioral break for already compiled dependents.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1515]: We also have the option of saying:

`DateTime` always overflows when used as a concrete type but when used in a generic math context it properly respects `checked`/`unchecked`. We can always have the public APIs and the explicit interface implementations SxS, just needing to have API review in agreement that's overall good/desirable

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1515]: Right.  Seems like we basically have four options:
1. Don't implement the generic math interfaces on these types.
2. Leave the behavior as is.  Someone using these types in a generic context might experience exceptions for "overflow" (but not actually OverflowException) when not in unchecked.
3. Implement the new unchecked operators on these types to not throw.  This means someone using these in a generic context will have more consistent behavior regardless of the concrete type being used, but it yields an inconsistency between how the type behaves when using it directly vs generically.
4. (3) and change the existing behavior of the existing operators to be non-throwing.  That's a breaking change in that someone could easily be relying on that behavior to stop "bad things" from happening.

We don't need to decide now, but please ensure there's an issue open to track making a decision around this for .NET 7.

**Review by stephentoub — APPROVED** [review]
Other than my question about using checked when it doesn't seem to have meaning, LGTM.

File: src/libraries/System.Private.CoreLib/src/System/Byte.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Byte.cs:395]: Right. For byte (and the other primitives) this is a case we'd switch over to a DIM once that's available. .NET only exposes a single instruction for these and so there's no need for the primitives to actually do anything special.

We likely still want to support the concept on the interface for end users who may want to provide their own overflow behavior if that makes sense for their domain (although my opinion here isn't very strong for division in particular).

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1515]: Will log an issue tracking this before merging.

File: src/libraries/System.Private.CoreLib/src/System/Byte.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Byte.cs:395]: > this is a case we'd switch over to a DIM once that's available

Cool.  Would it make sense then to remove `checked(...)` from this PR where it's not actually doing anything?  My reasoning is simply that it'll make it easier to spot all the operators we can remove once we have DIMs: if they're not using checked, remove them, if they are, keep them.  Up to you.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Byte.cs:395]: Can do. I made sure to remove them from float/single/double, DateTime, and TimeSpan; so it makes sense to do the same for integer division

File: src/libraries/System.Private.CoreLib/src/System/Byte.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Byte.cs:304]: These are explicitly implemented and so accessibility modifiers aren't allowed

File: src/libraries/System.Private.CoreLib/src/System/Byte.cs
Comment (wuhkuh) [inline src/libraries/System.Private.CoreLib/src/System/Byte.cs:414]: Out of curiosity: are the block comments excluded from the uncommenting pass purposely?

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Byte.cs:414]: Nope. That is a mistake and I'll fix it when I address the other relevant comments given above

File: src/libraries/System.Private.CoreLib/src/System/Byte.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Byte.cs:395]: Fixed.

File: src/libraries/System.Runtime/tests/System/ByteTests.GenericMath.cs
Comment (tannergooding) [reply src/libraries/System.Runtime/tests/System/ByteTests.GenericMath.cs:59]: The tests are explicitly simple and easily copyable here since I have to replicate it for each of many types.

These are basically just testing that the interface forwards down correctly and so don't need a lot of complexity.

File: src/libraries/System.Private.CoreLib/src/System/Byte.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Byte.cs:414]: Fixed.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1515]: https://github.com/dotnet/runtime/issues/67744

**Pull request merged**
